src/mip/utils/timestamp.hpp

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+#include <assert.h>
+#include <chrono>
+
+using std::chrono::duration_cast;
+using Nanoseconds = std::chrono::nanoseconds;
+using Microseconds = std::chrono::microseconds;
+using Milliseconds = std::chrono::milliseconds;
+using Seconds = std::chrono::seconds;
+using Minutes = std::chrono::minutes;
+using Hours = std::chrono::hours;
+
+#if _HAS_CXX20
+    using Days = std::chrono::days;
+    using Weeks = std::chrono::weeks;
+    using Months = std::chrono::months;
+    using Years = std::chrono::years;
+#else
+    #include <ratio>
+    using Days = std::chrono::duration<int, std::ratio<86400>>;
+    using Weeks = std::chrono::duration<int, std::ratio<604800>>;
+    using Months = std::chrono::duration<int, std::ratio<2629746>>;
+    using Years = std::chrono::duration<int, std::ratio<31556952>>;
+#endif // _HAS_CXX20
+
+namespace mip
+{
+    /// -----------------------------------------------------
+    /// Legend:
+    /// -----------------------------------------------------
+    /// T --> Arithmetic type
+    /// D --> Duration (currently std::chrono::duration only)
+    /// -----------------------------------------------------
+    
+
+    /// Manages a timestamp that increments based on an underlying time system.
+    ///
+    /// Notes:
+    /// ---------------------------------------------------------------------------------
+    ///     * Currently supports GPS clock (only) for the underlying time system.
+    ///     * Might add support for different time systems in the future.
+    ///     * Currently supports std::chrono:duration (only) for timestamp inputs.
+    /// ---------------------------------------------------------------------------------
+    class TimestampState
+    {
+    public:
+        TimestampState() {}
+
+        /// Synchronizes timestamp to the current time in the underlying time system.
+        void synchronize();
+
+        /// Returns time since epoch.
+        template<typename D> D getTimestamp();
+        Nanoseconds getTimestamp();
+
+    private:
+        Nanoseconds timestamp{0}; 
+    };
+
```

Comment:
All these template functions seem like good candidates to make member functions instead of always passing in a reference to the timestamp - is there a reason they're standalone?

Reply:
There is a reason they are standalone. So the member timestamp is actually different from the timestamps passed to the template functions (confusing I know). So that timestamp there holds the internal state of a TimestampState object, which is a timestamp referencing a coordinate time system (UTC, GPS, etc.). 

The functions were actually originally member functions. But now, the timestamp passed to those functions is actually an std::chrono::duration (for now). The idea here is that the template functions primarily handle manual timestamps (not referencing a time system directly, user entered, etc.), which can be implemented/stored any way developers need them to be. They just have to be able to pass the core timestamp duration to our functions. 

Some of the functions can be used by TimestampState objects too (such as getTimeOfWeek and timeChanged). But they don't only need to be used (and shouldn't need to be only used) by the TimestampState objects once again. So the TimestampState class provides getters for getting that std::chrono::duration timestamp representation that satisfies the generalized design of the functions.

Does that make sense? What do you think about that?

Reply:
Changed them to member functions.